Profiling can throw when using kNN with a pre-filter

### Elasticsearch Version

8.11+

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

_any_

### Problem Description

When profiling complex queries an exception can be thrown during profiling

### Steps to Reproduce

still working on this...

### Logs (if relevant)

```
Caused by: org.elasticsearch.search.dfs.DfsPhaseExecutionException: Dfs Failed [Exception during dfs phase]
  at org.elasticsearch.search.dfs.DfsPhase.execute(DfsPhase.java:61)
  at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:500)
  at org.elasticsearch.search.SearchService.lambda$executeDfsPhase$0(SearchService.java:489)
  at org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:51)
  at org.elasticsearch.action.ActionRunnable$2.accept(ActionRunnable.java:48)
  at org.elasticsearch.action.ActionRunnable$3.doRun(ActionRunnable.java:73)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:983)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  at java.lang.Thread.run(Thread.java:1583)
Caused by: org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper: index_out_of_bounds_exception: Index: 26, Size: 23
  at java.util.ArrayList.rangeCheckForAdd(ArrayList.java:838)
  at java.util.ArrayList.add(ArrayList.java:510)
  at org.elasticsearch.search.profile.AbstractInternalProfileTree.addDependencyNode(AbstractInternalProfileTree.java:103)
  at org.elasticsearch.search.profile.AbstractInternalProfileTree.getProfileBreakdown(AbstractInternalProfileTree.java:78)
  at org.elasticsearch.search.profile.AbstractProfiler.getQueryBreakdown(AbstractProfiler.java:26)
  at org.elasticsearch.search.internal.ContextIndexSearcher.createWeight(ContextIndexSearcher.java:225)
  at org.apache.lucene.search.BooleanWeight.<init>(BooleanWeight.java:59)
  at org.apache.lucene.search.BooleanQuery.createWeight(BooleanQuery.java:245)
  at org.elasticsearch.search.internal.ContextIndexSearcher.createWeight(ContextIndexSearcher.java:230)
  at org.apache.lucene.search.ConstantScoreQuery.createWeight(ConstantScoreQuery.java:136)
  at org.apache.lucene.search.AbstractMultiTermQueryConstantScoreWrapper$RewritingWeight.rewrite(AbstractMultiTermQueryConstantScoreWrapper.java:174)
  at org.apache.lucene.search.AbstractMultiTermQueryConstantScoreWrapper$RewritingWeight$1.get(AbstractMultiTermQueryConstantScoreWrapper.java:269)
  at org.elasticsearch.search.profile.query.ProfileWeight$1.get(ProfileWeight.java:69)
  at org.apache.lucene.search.Boolean2ScorerSupplier.req(Boolean2ScorerSupplier.java:212)
  at org.apache.lucene.search.Boolean2ScorerSupplier.getInternal(Boolean2ScorerSupplier.java:131)
  at org.apache.lucene.search.Boolean2ScorerSupplier.get(Boolean2ScorerSupplier.java:111)
  at org.elasticsearch.search.profile.query.ProfileWeight$1.get(ProfileWeight.java:69)
  at org.apache.lucene.search.Boolean2ScorerSupplier.req(Boolean2ScorerSupplier.java:212)
  at org.apache.lucene.search.Boolean2ScorerSupplier.getInternal(Boolean2ScorerSupplier.java:131)
  at org.apache.lucene.search.Boolean2ScorerSupplier.get(Boolean2ScorerSupplier.java:111)
  at org.elasticsearch.search.profile.query.ProfileWeight$1.get(ProfileWeight.java:69)
  at org.elasticsearch.search.profile.query.ProfileWeight.scorer(ProfileWeight.java:45)
  at org.apache.lucene.search.AbstractKnnVectorQuery.getLeafResults(AbstractKnnVectorQuery.java:134)
  at org.apache.lucene.search.AbstractKnnVectorQuery.searchLeaf(AbstractKnnVectorQuery.java:117)
  at org.apache.lucene.search.AbstractKnnVectorQuery.lambda$parallelSearch$0(AbstractKnnVectorQuery.java:111)
  at java.util.concurrent.FutureTask.run(FutureTask.java:317)
  ... 6 more
```

OK, I added a check to make sure the currentThread when constructing the profiler is the same thread when using it, and it turns out it isn't. I am guessing we haven't turned off multi-threading when profiling is turned on?

```
Caused by: java.lang.IllegalStateException: Profiler is not thread-safe and can only be used from the thread that created it. Current thread: elasticsearch[node_s0][search_worker][T#3], profiler thread: elasticsearch[node_s0][search][T#4]
```

Also note in my testing I periodically saw the following:
```
    Caused by: org.elasticsearch.search.dfs.DfsPhaseExecutionException: Dfs Failed [Exception during dfs phase]
        at org.elasticsearch.search.dfs.DfsPhase.execute(DfsPhase.java:62)
        at org.elasticsearch.search.SearchService.executeDfsPhase(SearchService.java:499)
        at org.elasticsearch.search.SearchService.lambda$executeDfsPhase$0(SearchService.java:488)
        at org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:73)
        at org.elasticsearch.action.ActionRunnable$3.accept(ActionRunnable.java:70)
        at org.elasticsearch.action.ActionRunnable$4.doRun(ActionRunnable.java:95)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
        at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base/java.lang.Thread.run(Thread.java:1583)
    Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because "parent" is null
        at org.elasticsearch.search.profile.AbstractInternalProfileTree.updateParent(AbstractInternalProfileTree.java:175)
        at org.elasticsearch.search.profile.AbstractInternalProfileTree.getProfileBreakdown(AbstractInternalProfileTree.java:72)
        at org.elasticsearch.search.profile.AbstractProfiler.getQueryBreakdown(AbstractProfiler.java:26)
        at org.elasticsearch.search.internal.ContextIndexSearcher.createWeight(ContextIndexSearcher.java:219)
        at org.apache.lucene.search.ConstantScoreQuery.createWeight(ConstantScoreQuery.java:136)
        at org.apache.lucene.search.AbstractMultiTermQueryConstantScoreWrapper$RewritingWeight.rewrite(AbstractMultiTermQueryConstantScoreWrapper.java:174)
        at org.apache.lucene.search.AbstractMultiTermQueryConstantScoreWrapper$RewritingWeight$1.get(AbstractMultiTermQueryConstantScoreWrapper.java:269)
        at org.elasticsearch.search.profile.query.ProfileWeight$1.get(ProfileWeight.java:69)
        at org.apache.lucene.search.Boolean2ScorerSupplier.req(Boolean2ScorerSupplier.java:212)
        at org.apache.lucene.search.Boolean2ScorerSupplier.getInternal(Boolean2ScorerSupplier.java:131)
        at org.apache.lucene.search.Boolean2ScorerSupplier.get(Boolean2ScorerSupplier.java:111)
        at org.elasticsearch.search.profile.query.ProfileWeight$1.get(ProfileWeight.java:69)
        at org.elasticsearch.search.profile.query.ProfileWeight.scorer(ProfileWeight.java:45)
        at org.apache.lucene.search.AbstractKnnVectorQuery.getLeafResults(AbstractKnnVectorQuery.java:116)
        at org.apache.lucene.search.AbstractKnnVectorQuery.searchLeaf(AbstractKnnVectorQuery.java:99)
        at org.apache.lucene.search.AbstractKnnVectorQuery.lambda$rewrite$0(AbstractKnnVectorQuery.java:86)
        at org.apache.lucene.search.TaskExecutor$TaskGroup.lambda$createTask$0(TaskExecutor.java:118)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
        ... 6 more
    }
        at __randomizedtesting.SeedInfo.seed([F5E517678C2E344F:6537B3C296313554]:0)
        at app//org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:708)
        at app//org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:400)
        at app//org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:740)
        at app//org.elasticsearch.action.search.AbstractSearchAsyncAction.onShardFailure(AbstractSearchAsyncAction.java:493)
        at app//org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:331)
        at app//org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:54)
        at app//org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:656)
        at app//org.elasticsearch.transport.TransportService$UnregisterChildTransportResponseHandler.handleException(TransportService.java:1717)
        at app//org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1435)
        at app//org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1575)
        at app//org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1550)
        at app//org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:46)
        at app//org.elasticsearch.action.support.ChannelActionListener.onFailure(ChannelActionListener.java:38)
        at app//org.elasticsearch.action.ActionRunnable.onFailure(ActionRunnable.java:146)
        at app//org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
        at app//org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:33)
        at app//org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:984)
        at app//org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
        at java.base@21/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
        at java.base@21/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
        at java.base@21/java.lang.Thread.run(Thread.java:1583)

        Caused by:
        org.elasticsearch.ElasticsearchException$1: Cannot invoke "java.lang.Integer.intValue()" because "parent" is null
            at app//org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:673)
            at app//org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:398)
            ... 19 more

            Caused by:
            java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because "parent" is null
                at org.elasticsearch.search.profile.AbstractInternalProfileTree.updateParent(AbstractInternalProfileTree.java:175)
                at org.elasticsearch.search.profile.AbstractInternalProfileTree.getProfileBreakdown(AbstractInternalProfileTree.java:72)
                at org.elasticsearch.search.profile.AbstractProfiler.getQueryBreakdown(AbstractProfiler.java:26)
                at org.elasticsearch.search.internal.ContextIndexSearcher.createWeight(ContextIndexSearcher.java:219)
                at org.apache.lucene.search.ConstantScoreQuery.createWeight(ConstantScoreQuery.java:136)
                at org.apache.lucene.search.AbstractMultiTermQueryConstantScoreWrapper$RewritingWeight.rewrite(AbstractMultiTermQueryConstantScoreWrapper.java:174)
                at org.apache.lucene.search.AbstractMultiTermQueryConstantScoreWrapper$RewritingWeight$1.get(AbstractMultiTermQueryConstantScoreWrapper.java:269)
                at org.elasticsearch.search.profile.query.ProfileWeight$1.get(ProfileWeight.java:69)
                at org.apache.lucene.search.Boolean2ScorerSupplier.req(Boolean2ScorerSupplier.java:212)
                at org.apache.lucene.search.Boolean2ScorerSupplier.getInternal(Boolean2ScorerSupplier.java:131)
                at org.apache.lucene.search.Boolean2ScorerSupplier.get(Boolean2ScorerSupplier.java:111)
                at org.elasticsearch.search.profile.query.ProfileWeight$1.get(ProfileWeight.java:69)
                at org.elasticsearch.search.profile.query.ProfileWeight.scorer(ProfileWeight.java:45)
                at org.apache.lucene.search.AbstractKnnVectorQuery.getLeafResults(AbstractKnnVectorQuery.java:116)
                at org.apache.lucene.search.AbstractKnnVectorQuery.searchLeaf(AbstractKnnVectorQuery.java:99)
                at org.apache.lucene.search.AbstractKnnVectorQuery.lambda$rewrite$0(AbstractKnnVectorQuery.java:86)
                at org.apache.lucene.search.TaskExecutor$TaskGroup.lambda$createTask$0(TaskExecutor.java:118)
                at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
                ... 6 more
```

--- a/docs/changelog/108041.yaml
+++ b/docs/changelog/108041.yaml
@@ -1,0 +1,7 @@
+pr: 108041
+summary: Handle parallel calls to `createWeight` when profiling is on
+area: Search
+type: bug
+issues:
+ - 104131
+ - 104235

--- a/server/src/internalClusterTest/java/org/elasticsearch/search/profile/dfs/DfsProfilerIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/search/profile/dfs/DfsProfilerIT.java
@@ -39,7 +39,6 @@ public class DfsProfilerIT extends ESIntegTestCase {
 
     private static final int KNN_DIM = 3;
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/104235")
     public void testProfileDfs() throws Exception {
         String textField = "text_field";
         String numericField = "number";

--- a/server/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java
+++ b/server/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java
@@ -48,7 +48,7 @@ public abstract class AbstractInternalProfileTree<PB extends AbstractProfileBrea
      * @param query The scoring query we wish to profile
      * @return      A ProfileBreakdown for this query
      */
-    public PB getProfileBreakdown(E query) {
+    public final synchronized PB getProfileBreakdown(E query) {
         int token = currentToken;
 
         boolean stackEmpty = stack.isEmpty();
@@ -109,7 +109,7 @@ public abstract class AbstractInternalProfileTree<PB extends AbstractProfileBrea
     /**
      * Removes the last (e.g. most recent) value on the stack
      */
-    public void pollLast() {
+    public final synchronized void pollLast() {
         stack.pollLast();
     }
 
@@ -120,7 +120,7 @@ public abstract class AbstractInternalProfileTree<PB extends AbstractProfileBrea
      *
      * @return a hierarchical representation of the profiled query tree
      */
-    public List<ProfileResult> getTree() {
+    public final synchronized List<ProfileResult> getTree() {
         ArrayList<ProfileResult> results = new ArrayList<>(roots.size());
         for (Integer root : roots) {
             results.add(doGetTree(root));

--- a/server/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java
+++ b/server/src/main/java/org/elasticsearch/search/profile/AbstractInternalProfileTree.java
@@ -18,24 +18,16 @@ import java.util.List;
 
 public abstract class AbstractInternalProfileTree<PB extends AbstractProfileBreakdown<?>, E> {
 
-    protected ArrayList<PB> breakdowns;
+    private final ArrayList<PB> breakdowns = new ArrayList<>(10);
     /** Maps the Query to it's list of children.  This is basically the dependency tree */
-    protected ArrayList<ArrayList<Integer>> tree;
+    private final ArrayList<ArrayList<Integer>> tree = new ArrayList<>(10);
     /** A list of the original queries, keyed by index position */
-    protected ArrayList<E> elements;
+    private final ArrayList<E> elements = new ArrayList<>(10);
     /** A list of top-level "roots".  Each root can have its own tree of profiles */
-    protected ArrayList<Integer> roots;
+    private final ArrayList<Integer> roots = new ArrayList<>(10);
     /** A temporary stack used to record where we are in the dependency tree. */
-    protected Deque<Integer> stack;
+    private final Deque<Integer> stack = new ArrayDeque<>(10);
     private int currentToken = 0;
-
-    public AbstractInternalProfileTree() {
-        breakdowns = new ArrayList<>(10);
-        stack = new ArrayDeque<>(10);
-        tree = new ArrayList<>(10);
-        elements = new ArrayList<>(10);
-        roots = new ArrayList<>(10);
-    }
 
     /**
      * Returns a {@link QueryProfileBreakdown} for a scoring query.  Scoring queries (e.g. those

--- a/server/src/main/java/org/elasticsearch/search/profile/AbstractProfileBreakdown.java
+++ b/server/src/main/java/org/elasticsearch/search/profile/AbstractProfileBreakdown.java
@@ -44,7 +44,7 @@ public abstract class AbstractProfileBreakdown<T extends Enum<T>> {
      * @param timingType the timing type to create a new {@link Timer} for
      * @return a new {@link Timer} instance
      */
-    public Timer getNewTimer(T timingType) {
+    public final Timer getNewTimer(T timingType) {
         Timer timer = new Timer();
         timings.get(timingType).add(timer);
         return timer;

